hal/mbed_ticker_api.c

```diff
@@ -13,61 +13,178 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+ #include <stdio.h>
 #include <stddef.h>
 #include "hal/ticker_api.h"
 #include "platform/mbed_critical.h"
 
-void ticker_set_handler(const ticker_data_t *const data, ticker_event_handler handler) {
-    data->interface->init();
+#define MBED_MIN(x,y) (((x)<(y))?(x):(y))
 
-    data->queue->event_handler = handler;
+static void update_interrupt(const ticker_data_t *const ticker);
+static void update_current_timestamp(const ticker_data_t *const ticker);
+
+/*
+ * Initialize a ticker instance.  
+ */
+static void initialize(const ticker_data_t *ticker) { 
+    // return if the queue has already been initialized, in that case the 
+    // interface used by the queue is already initialized.
+    if (ticker->queue->initialized) { 
+        return;
+    }
+
+    ticker->interface->init();
+    
+    ticker->queue->event_handler = NULL;
+    ticker->queue->head = NULL;
+    ticker->queue->timestamp = 0;
+    ticker->queue->initialized = true;
+    
+    update_current_timestamp(ticker);
+    update_interrupt(ticker);
+}
+
+/**
+ * Set the event handler function of a ticker instance. 
+ */
+static void set_handler(const ticker_data_t *const ticker, ticker_event_handler handler) { 
+    ticker->queue->event_handler = handler;
+}
+
+/*
+ * Convert a low res timestamp to a high res timestamp. An high resolution 
+ * timestamp is used as the reference point to convert the low res timestamp 
+ * into an high res one. 
+ * 
+ * It is important to note that the result will **never** be in the past. If the 
+ * value of the low res timetamp is less than the low res part of the reference 
+ * timestamp then an overflow is 
+ * 
+ * @param ref: The timestamp of reference. 
+ * @param relative_timestamp: The timestamp to convert. 
+ */
+static us_timestamp_t convert_relative_timestamp(us_timestamp_t ref, timestamp_t relative_timestamp) { 
+     bool overflow = relative_timestamp < ((timestamp_t) ref) ? true : false;
+
+    us_timestamp_t result = (ref & ~((us_timestamp_t)UINT32_MAX)) | relative_timestamp;
+    if (overflow) { 
+        result += (1ULL<<32);
+    }
+
+    return result;
+}
+
+/**
+ * update the current timestamp value of a ticker.
```

Comment:
User is not supposed to call this function, it is private (static). It can be called whenever the us_timestamp_t held by a `ticker_data_t` has to be updated.